libs/bytecode/src/utils.sw

```diff
@@ -0,0 +1,189 @@
+        if size == 1 {
+            break;
+        }
+
+        odd = (size & 1);
+        size = (size + 1) >> 1;
+    }
+
+    vec_digest.get(0).unwrap()
+}
+
+/// Swaps out configurable values in a contract or predicate's bytecode.
+pub fn _swap_configurables(bytecode: Vec<u8>, configurables: Vec<(u64, Vec<u8>)>) -> Vec<u8> {
+    // Copy the bytecode to a newly allocated memory to avoid memory ownership error.
+    let mut bytecode_slice = raw_slice::from_parts::<u8>(alloc_bytes(bytecode.len()), bytecode.len());
+    bytecode
+        .buf
+        .ptr
+        .copy_bytes_to(bytecode_slice.ptr(), bytecode.len());
+
+    // Iterate over every configurable
+    let mut configurable_iterator = 0;
+    while configurable_iterator < configurables.len() {
+        let (offset, data) = configurables.get(configurable_iterator).unwrap();
+
+        // Overwrite the configurable data into the bytecode
+        data.buf
+            .ptr
+            .copy_bytes_to(bytecode_slice.ptr().add::<u8>(offset), data.len());
+
+        configurable_iterator += 1;
+    }
+
+    Vec::from(bytecode_slice)
+}
+
+/// Takes some bytes and creates a new leaf digest.
+fn leaf_digest(data: raw_slice) -> b256 {
+    let number_of_bytes = data.number_of_bytes();
+    let mut bytes = Bytes::with_capacity(number_of_bytes + 1);
+
+    // Prepend LEAF to the leaf bytes
+    bytes.buf.ptr().write_byte(LEAF);
+    data
+        .ptr()
+        .copy_bytes_to(bytes.buf.ptr().add_uint_offset(1), number_of_bytes);
+    bytes.len = number_of_bytes + 1;
+
+    // Compute the digest
+    let mut result_buffer = b256::min();
+    asm(hash: result_buffer, ptr: bytes.buf.ptr, bytes: bytes.len) {
+        s256 hash ptr bytes;
+        hash: b256
+    }
+}
+
+/// Takes two nodes and creates a new node digest.
+fn node_digest(left: b256, right: b256) -> b256 {
+    let mut bytes = Bytes::with_capacity(65);
+    let new_ptr_left = bytes.buf.ptr().add_uint_offset(1);
+    let new_ptr_right = bytes.buf.ptr().add_uint_offset(33);
+
+    // Prepend NODE and concat the 2 node bytes for the current node
+    bytes.buf.ptr().write_byte(NODE);
+    __addr_of(left).copy_bytes_to(new_ptr_left, 32);
+    __addr_of(right).copy_bytes_to(new_ptr_right, 32);
+    bytes.len = 65;
+
+    // Compute the digest
+    let mut result_buffer = b256::min();
+    asm(hash: result_buffer, ptr: bytes.buf.ptr, bytes: bytes.len) {
+        s256 hash ptr bytes;
+        hash: b256
+    }
+}
+
+/// Takes some bytecode and computes the resulting leaves for a merkle tree.
+fn generate_leaves(bytecode: Vec<u8>) -> Vec<b256> {
+    // Number of leaves is '(bytecode.len() / LEAF_SIZE)' if it's perfectly divisible by LEAF_SIZE.
+    // Otherwise `(bytecode.len() / LEAF_SIZE) + 1` to account for padding.
+    let num_leaves = match bytecode.len() % LEAF_SIZE {
+        0 => bytecode.len() / LEAF_SIZE,
+        _ => bytecode.len() / LEAF_SIZE + 1,
+    };
+    let mut leaves_digest: Vec<b256> = Vec::new();
```

Comment:
Can this be switched to `with_capacity()` using `num_leaves`?

Reply:
This has been converted into a `raw_slice` in my recent optimization commit